File: 100daysofcode/priority_queue_using_min_binary_heap.cpp
#include<iostream>
#include<vector>
#include<math.h>
using namespace std;

struct Node{
    int val,priority;
    Node(int val,int priority){
        this->val=val;
        this->priority=priority;
    }
};

struct priorityQueue{
    vector<struct Node*> values;
    void enqueue(int val,int priority){
        Node *newnode=new Node(val,priority);
        values.push_back(newnode);
        bubbleup();
    }
    void bubbleup(){
        int idx = values.size()-1;
        struct Node* element=values[idx];
        while(idx>0){
            int parentidx = floor((idx-1)/2);
            struct Node* parent =values[parentidx];
            if(parent->priority<=element->priority) break;
            values[parentidx]=element;
            values[idx]=parent;
            idx=parentidx;
        }
    }
    struct Node* dequeue(){
		struct Node* min =values[0];
		struct Node* end =values.back();
        values.pop_back();
		if(values.size()>0){
			values[0]=end;
		    sinkDown();	
		}
		return min;
	}
	void sinkDown(){
		int idx=0;
		int length =values.size();
		struct Node* element = values[0];
		while(true){
			int leftChildIdx =2*idx+1;
			int rightChildIdx =2*idx+2;
			struct Node *leftChild,*rightChild;
			bool swap=false;
            int s;
			if(leftChildIdx<length){
				leftChild=values[leftChildIdx];
				if(leftChild->priority < element->priority){
					s=leftChildIdx;
                    swap=true;
				}
			}
			if(rightChildIdx<length){
				rightChild=values[rightChildIdx];
				if((swap==false && rightChild->priority <element->priority)||
				(swap!=false &&rightChild->priority < leftChild->priority)){
					s=rightChildIdx;
                    swap=true;
				}
			}
			if(swap==false) break;
			values[idx]=values[s];
			values[s]=element;
			idx=s;
		}
	}
    void display(){
        cout<<"The priority min binary heap:\n";
        for(int i=0;i<values.size();i++){
            cout<<"priority: "<<values[i]->priority<<" "<<"value: "<<values[i]->val<<endl;
        }
    }
};

int main(){
    struct priorityQueue pq;
    pq.enqueue(10,4);pq.enqueue(100,5);
    pq.enqueue(200,1);pq.enqueue(25,3);
    pq.enqueue(46,0);pq.enqueue(150,2);
    pq.display();
    cout<<"\ndequeued: value: "<<pq.dequeue()->val<<endl;
    pq.display();
    cout<<"\ndequeued: value: "<<pq.dequeue()->val<<endl;
    pq.display();
    cout<<"\ndequeued: value: "<<pq.dequeue()->val<<endl;
    pq.display();
    cout<<"\ndequeued: value: "<<pq.dequeue()->val<<endl;
    pq.display();
    cout<<"\ndequeued: value: "<<pq.dequeue()->val<<endl;
    pq.display();
    return 0;
}

File: binary_heap/priority_queue_using_min_binary_heap.js
class priorityQueue{
	constructor(){
		this.values = [];
	}
	
	enqueue(val,priority){
	    let newnode =new Node(val,priority);
		this.values.push(newnode);
		this.bubbleup();
	}
	bubbleup(){
		let idx = this.values.length - 1;
		const element = this.values[idx];
		while(idx > 0){
			let parentidx = Math.floor((idx-1)/2);
			let parent = this.values[parentidx];
			if(element.priority >= parent.priority)break;
			this.values[parentidx] = element;
			this.values[idx]=parent;
			idx = parentidx;
		}
	}

	dequeue(){
		const min =this.values[0];
		const end =this.values.pop();
		if(this.values.length>0){
			this.values[0]=end;
		    this.sinkDown();	
		}
		return min;
	}
	sinkDown(){
		let idx=0;
		const length =this.values.length;
		const element = this.values[0];
		while(true){
			let leftChildIdx =2*idx+1;
			let rightChildIdx =2*idx+2;
			let leftChild,rightChild;
			let swap=null;
			if(leftChildIdx<length){
				leftChild=this.values[leftChildIdx];
				if(leftChild.priority<element.priority){
					swap=leftChildIdx;
				}
			}
			if(rightChildIdx<length){
				rightChild=this.values[rightChildIdx];
				if((swap===null && rightChild.priority<element.priority)||
				(swap!==null &&rightChild.priority<leftChild.priority)){
					swap=rightChildIdx;
				}
			}
			if(swap===null) break;
			this.values[idx]=this.values[swap];
			this.values[swap]=element;
			idx=swap;
		}
	}
}
class Node{
    constructor(val,priority){
        this.val=val;
        this.priority =priority;
    }
}

let er =new priorityQueue();
er.enqueue("common cold",5)
er.enqueue("gunshot wound",1)
er.enqueue("high fever",4)
er.enqueue("broken arm",2)
er.enqueue("glass cut",3)